src/Microsoft.AspNetCore.Mvc.Core/Routing/UrlHelper.cs

```diff
@@ -233,6 +233,13 @@ protected virtual string GenerateUrl(string protocol, string host, VirtualPathDa
             // VirtualPathData.VirtualPath returns string.Empty instead of null.
             Debug.Assert(pathData.VirtualPath != null);
 
+            var url = FastGetUrl(protocol, host, pathData, fragment);
+
+            if (!string.IsNullOrEmpty(url))
```

Comment:
Ryan, what about a `bool TryFastGenerateUrl` ? Or anything that returns a bool to check the result.